index.d.ts

```diff
@@ -0,0 +1,60 @@
+export interface Action {
+  type: string;
+}
+
+export type Reducer<S> = (state: S, action: Action) => S;
+
+export type Dispatch = (action: any) => any;
+
+export interface MiddlewareAPI<S> {
+  dispatch: Dispatch;
+  getState: () => S;
+}
+
+export interface Middleware {
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch) => Dispatch;
+}
```

Comment:
Middleware definition below seems to be enough, but…

```typescript
interface Middleware<S> extends Function {
    (store: Store<S>): (next: Dispatch) => Dispatch;
}
```

It actually does not return a dispatch but function mapping input of type A to output of type B:

```typescript
interface Middleware<S, A, B> extends Function {
    (store: Store<S>): (next: Dispatch) => (action: A) => B;
}
```

But in this case `Middleware` will always have to be parametrised with both type parameters. We can avoid this ( *but I'm NOT sure if we should* ) in the same manner as with Dispatch:

```typescript
interface Middleware extends Function {
    <S, A, B>(store: Store<S>): (next: Dispatch) => (action: A) => B;
}
```

It's not always so easy to add static type definitions to code written in dynamically typed language… ;)